src/control/pbin/exec.go

```diff
@@ -150,25 +188,29 @@ func ExecReq(parent context.Context, log logging.Logger, binPath string, req *Re
 	if _, err := conn.Write(sendData); err != nil {
 		return nil, errors.Wrap(err, "pbin write failed")
 	}
+	// Signal to the receiver that we're finished.
+	if err := conn.CloseWrite(); err != nil {
+		return nil, errors.Wrap(err, "pbin CloseWrite failed")
+	}
 
-	maxReadAttempts := 5
 	readCount := 0
+	maxReadAttempts := 5
 	for {
-		recvData := make([]byte, MaxMessageSize)
-		recvLen, err := conn.Read(recvData)
+		recvData, err := ReadMessage(conn)
 		if err != nil {
-			return nil, errors.Wrap(err, "pbin read failed")
+			return nil, errors.Wrap(err, "failed to read message from sender")
 		}
 
-		res, err = decodeResponse(recvData[:recvLen])
+		res, err = decodeResponse(recvData)
 		if err != nil && err != io.EOF {
-			log.Debugf("discarding garbage response %q", recvData[:recvLen])
-			readCount++
+			log.Debugf("discarding garbage response %q", recvData)
 			if readCount < maxReadAttempts {
+				readCount++
```

Comment:
If you move readCount++ here, then it allows maxReadAttempts + 1 reads.  In this case, you'd allow 6, when maxReadAttemps said 5.

Reply:
D'oh, you're right. When I was refactoring I moved where the increment happens and didn't notice this.